### ISSUE TITLE
Adding entropy function analogous to SciPy

### ISSUE THREAD

**Issue by BIT-silence** [pytorch/pytorch#43255]
## 🚀 Feature
Add an entropy function in PyTorch to compute entropy = -sum(p * log(p), axis=axis), which is analogous to scipy.stats.entropy.

## Motivation

It is a common thing to compute entropy given a distribution p in many use cases. Currently there are several ways to compute entropy in PyTorch listed as below.

1. Direct implementation which is `entropy = -((p * p.log()).sum(dim=-1))`. This one works in many cases, but not elegant and efficient enough.
2. Using torch.bmm as `entropy = -(torch.bmm(p.view(M, 1, N), p.log().view(M, N, 1))).squeeze()`. This one have better forward pass performance compared to the 1st one on CPU when MKL is enabled. However, the backward pass performance will be worse because of the backward computation of bmm.
3. Using torch.distributions.categorical.Categorical.entropy as `entropy = torch.distributions.categorical.Categorical(p).entropy()`. This one looks the most direct one. However, it is much slower than both of the previous approaches in both forward and backward pass even if we don't need logit at all.

The 1st and 2nd approaches are not a direct function to compute entropy as scipy. So we can consider either to add a scipy-like entropy function or optimize the implementation of torch.distributions.categorical.Categorical.entropy.

**Comment by BIT-silence** [comment]
cc @mruberry @ngimel

**Comment by vadimkantorov** [comment]
Quite related: https://github.com/pytorch/pytorch/issues/9993 https://github.com/pytorch/pytorch/issues/22656 https://github.com/pytorch/pytorch/pull/40336#issuecomment-646999051

**Comment by vadimkantorov** [comment]
`torch.special.entr()` was added in https://github.com/pytorch/pytorch/pull/53500

The differnece from `entropy` would be the sum-reduction/averaging across a dim?
